Fix build with boost ASIO 1.87. Support boost 1.66+

vtnerd opened the pull request:
This should compile with Boost 1.73-1.87. I have only tested on Boost 1.74 and Boost 1.87 (primarily unit tests). I will mark the places where I know changes need to be made for older boost versions. I will clear up all such errors by manually downloading Boost 1.66 and testing.

One problem I have with supporting Boost 1.66-1.69 is that most types require an io_context object, whereas Boost 1.70+ have relaxed this requirement to an executor object. I will show an example of where it matters; it will be easy to break older Boost versions unintentionally.

This is currently marked as a draft, because of the items pointing out in a subsequent review.

vtnerd commented on contrib/epee/include/net/abstract_tcp_server2.inl:
```diff
@@ -1194,79 +1198,71 @@ namespace net_utils
+        MERROR("Failed to bind IPv6: " << e.what());
+        throw;
       }
     }
```
There should be an `else` here that logs an an error and returns `false`.

vtnerd replied:
Reverted to previous structure.

vtnerd commented on contrib/epee/src/net_ssl.cpp:
```diff
@@ -527,7 +528,7 @@ void ssl_options_t::configure(
       // when fingerprints are whitelisted.
       const bool verified = preverified &&
-        (verification != ssl_verification_t::system_ca || host.empty() || boost::asio::ssl::rfc2818_verification(host)(preverified, ctx));
+        (verification != ssl_verification_t::system_ca || host.empty() || boost::asio::ssl::host_name_verification(host)(preverified, ctx));
```
This change should fail on Boost 1.66-1.73. A `#ifdef` will be needed to differentiate between the two names.

vtnerd replied:
Fixed.

vtnerd commented on contrib/epee/src/net_helper.cpp:
```diff
@@ -52,12 +23,12 @@ namespace net_utils
 			boost::asio::ip::tcp::socket socket_;
 
-			explicit new_connection(boost::asio::io_service& io_service)
+			explicit new_connection(const boost::asio::steady_timer::executor_type& io_service)
```
This should fail on the `socket_` constructor on Boost 1.66-1.69. The problem is that `socket_` will require an `io_context` object, whereas newer Boost versions only need an `executor`. This will likely be the source of many compile errors moving forward.

I'm not certain how I will fix this, I may have to change this function signature which may break some downstream projects.

vtnerd replied:
Fixed.

tobtoht commented:
(This needs #9162 to pass depends CI. I will undraft that PR after the next round of merges.)

vtnerd commented:
The latest force push should work with Boost 1.66-1.87. Something funky happened with the tests, so I may push a dummy change later to see what happens.

tobtoht commented:
https://github.com/tobtoht/monero/actions/runs/12390979989/job/34587144842#step:5:1281

Looks like `boost::asio::io_service` is also used in `src/device_trezor/trezor`.

jeffro256 commented on contrib/epee/include/net/abstract_tcp_server2.h:
```diff
@@ -267,13 +268,14 @@ namespace net_utils
-    /// Construct a connection with the given io_service.
-    explicit connection( boost::asio::io_service& io_service,
+    /// Construct a connection with the given io_context.
+    explicit connection( boost::asio::io_context& io_context,
```
Shouldn't we use the `io_context_t` alias here since we're changing this line anyways?

vtnerd replied:
Updated.

vtnerd commented:
Fore pushed some changes:
  * Reverted changes to query code, they are basically same style with new interface
  * Switched to `io_context_t` as @jeffro256 suggested
  * Fix Trezor build.

vtnerd commented:
Force pushed twice with updates:

  * Added a missing include, but it doesn't matter as the depends builds is using Boost 1.64 which won't work
  * Reverted some whitespace changes that were made when making the patch.

vtnerd commented:
Force pushed an include fix.

vtnerd commented:
This patch looks ready. The depends CI will all fail until #9162 gets merged. I will leave as draft for now for that reason.

selsta reviewed (commented):
README needs an update to indicate the new minimum version

0xFFFC0000 commented:
Once this is ready for review, please remove Draft status.

vtnerd commented:
Force pushed:
  - A rebase of recent `master` merges.
  - Reverted #9625 (no longer needed)
  - Reverted #9623 (no longer needed)
  - Requires #9162 for depends CI to pass

Likely this will get removed as draft, assuming some of the tests cycle green.

vtnerd commented:
> README needs an update to indicate the new minimum version

Forgot to update this in the last force push, so another quick push with this change.

vtnerd commented:
Now ready for review.

jeffro256 commented:
Getting a ton of build errors across many platforms:

```
fatal error: boost/asio/io_context.hpp: No such file or directory
```

jeffro256 commented:
The dependencies system seems to be pulling Boost 1.64 instead of 1.66 on these failed platforms

selsta commented:
@jeffro256 requires #9162 to be merged first

jeffro256 commented:
Sorry, my bad. Should've read farther up in the comments

0xFFFC0000 reviewed (commented):
From technical perspective, everything is great, and I don't see any specific issues. 

The only remaining thing is I have to fetch #9162 and compile, and run all the tests.

jeffro256 commented on contrib/epee/src/connection_basic.cpp:
```diff
@@ -127,21 +121,21 @@ connection_basic_pimpl::connection_basic_pimpl(const std::string &name) : m_thro
 {
 	// add nullptr checks if removed
-	assert(m_state != nullptr); // release runtime check in get_context
+	assert(m_state != nullptr); // release runtime check in get_contextp
```
typo?

vtnerd replied:
Yup, fixed.

jeffro256 commented on contrib/epee/src/connection_basic.cpp:
```diff
@@ -127,21 +121,21 @@ connection_basic_pimpl::connection_basic_pimpl(const std::string &name) : m_thro
+	assert(m_state != nullptr); // release runtime check in get_contextp
 
-        socket_.next_layer() = std::move(sock);
+  socket_.next_layer() = std::move(sock);  
```
what's happening with the whitespace here?

vtnerd replied:
I'm not sure why I ever changed this, but this should be reverted now.

jeffro256 commented on contrib/epee/src/net_helper.cpp:
```diff
@@ -37,27 +53,20 @@ namespace net_utils
+			results = resolver.resolve(
+				boost::asio::ip::tcp::v6(), addr, port, boost::asio::ip::tcp::resolver::canonical_name
+			);
+			if (results.empty())
```
I think that this overload of `resolve()` shouldn't ever return an empty range because otherwise, it will throw a boost exception. That would make this `if` branch unreachable.

vtnerd replied:
Again, I was trying to minimize unrelated changes in this PR. I'd rather this be done independently and tested, etc.

jeffro256 commented on contrib/epee/src/net_helper.cpp:
```diff
@@ -37,27 +53,20 @@ namespace net_utils
 
 		//////////////////////////////////////////////////////////////////////////
 
-		struct new_connection
```
Was the move of this struct declaration purely for readability purposes?

vtnerd replied:
No, the `struct` needed a templated parameter because the type is `boost::asio::io_context` or `boost::asio::executor` depending on the Boost version. The compiler complained about templated functions inside of a function declaration, so I moved it.

jeffro256 reviewed (changes requested):
A couple actual comments, but mostly nitpicks

vtnerd commented on contrib/epee/include/net/abstract_tcp_server2.h:
```diff
@@ -100,8 +101,8 @@ namespace net_utils
-    using io_context_t = boost::asio::io_service;
-    using strand_t = boost::asio::io_service::strand;
+    using io_context_t = boost::asio::io_context;
+    using strand_t = boost::asio::io_context::strand;
```
Updated.

vtnerd commented on contrib/epee/include/net/abstract_tcp_server2.h:
```diff
@@ -306,7 +308,7 @@ namespace net_utils
     virtual bool call_run_once_service_io();
     virtual bool request_callback();
-    virtual boost::asio::io_service& get_io_service();
+    virtual boost::asio::io_context& get_io_context();
```
Updated.

vtnerd commented on contrib/epee/include/net/abstract_tcp_server2.inl:
```diff
@@ -1767,7 +1776,7 @@ namespace net_utils
     {
       new_connection_l->get_context(conn_context);
-      //new_connection_l.reset(new connection<t_protocol_handler>(io_service_, m_config, m_sock_count, m_pfilter));
+      //new_connection_l.reset(new connection<t_protocol_handler>(io_context_, m_config, m_sock_count, m_pfilter));
```
Done.

vtnerd commented on contrib/epee/include/net/abstract_tcp_server2.inl:
```diff
@@ -1796,14 +1805,16 @@ namespace net_utils
     {
       //resolving ipv4 address as ipv6 throws, catch here and move on
-      iterator = resolver.resolve(query, resolve_error);
+      results = resolver.resolve(
```
I initially started to do cleanup in some of these DNS queries, but then backed them all out because I think it's more appropriate in a separate patch. That way the commits-per-line match with a better explanation, etc.

vtnerd commented on tests/net_load_tests/net_load_tests.h:
```diff
@@ -32,7 +32,7 @@
 #include <atomic>
 
-#include <boost/asio/io_service.hpp>
+#include <boost/asio/io_context.hpp>
```
Doesn't look like its needed, removed.

vtnerd commented on contrib/epee/include/net/abstract_tcp_server2.inl:
```diff
@@ -1671,14 +1677,16 @@ namespace net_utils
     {
       //resolving ipv4 address as ipv6 throws, catch here and move on
-      iterator = resolver.resolve(query, resolve_error);
+      results = resolver.resolve(
```
Same answer as other - I'd rather do this in a separate PR.

vtnerd commented:
@selsta should this be done on the release branch? I don't remember what was decided.

selsta commented:
@vtnerd yes, please also open against release
